scripts/processors/planner-pipette.lua

```diff
@@ -0,0 +1,189 @@
+    end
+    cursor.set_entity_filter(#cursor.entity_filters+1, prototype)
+    return true
+
+  elseif cursor.is_upgrade_item then
+    local next_upgrade = prototype.next_upgrade
+    local to = next_upgrade and next_upgrade.name or global.downgrades[prototype.name]
+    return add_upgrade_filter(cursor, prototype.name, to)
+  end
+end
+
+script.on_event("bpt-pipette-add", function(event)
+  local cursor = get_cursor_stack(event.player_index)
+  if not cursor then return end
+  local prototype = get_selected_prototype(event)
+  if not prototype then return end
+
+  local player = game.get_player(event.player_index) --[[@as LuaPlayer]]
+  if add_filter(cursor, prototype) then
+    player.play_sound{path = "utility/smart_pipette"}
+  else
+    util.cursor_notification(player, { "message.bpt-unable-to-add-filter" }, "utility/cannot_build")
+  end
+end)
+
+---@param cursor LuaItemStack
+---@param prototype LuaEntityPrototype
+---@return boolean?
+local function remove_filter(cursor, prototype)
+  if cursor.is_deconstruction_item then
+    local filters = cursor.entity_filters
+    for i = 1, cursor.entity_filter_count do
+      if filters[i] == prototype.name then
+        cursor.set_entity_filter(i, nil)
+        return true
+      end
+    end
+
+  elseif cursor.is_upgrade_item then
+    for i = 1, cursor.prototype.mapper_count do
+      local mapper = cursor.get_mapper(i, "from")
+      if mapper.type == "entity" and mapper.name == prototype.name then
+        cursor.set_mapper(i, "from", nil)
+        cursor.set_mapper(i, "to", nil)
+        return true
+      end
+    end
+  end
+end
+
+script.on_event("bpt-pipette-remove", function(event)
+  local cursor = get_cursor_stack(event.player_index)
+  if not cursor then return end
+  local prototype = get_selected_prototype(event)
+  if not prototype then return end
+
+  local player = game.get_player(event.player_index) --[[@as LuaPlayer]]
+  if remove_filter(cursor, prototype) then
+    player.play_sound{path = "utility/clear_cursor"}
+  else
+    util.cursor_notification(player, { "message.bpt-unable-to-remove-filter" }, "utility/cannot_build")
+  end
+end)
+
+script.on_event("bpt-pipette-downgrade", function(event)
+  local cursor = get_cursor_stack(event.player_index)
+  if not cursor then return end
+  if not cursor.is_upgrade_item then return end
+
+  local prototype = get_selected_prototype(event)
+  if not prototype then return end
+
+  local downgrade = global.downgrades[prototype.name]
+  local next_upgrade = prototype.next_upgrade
+  local to = downgrade or (next_upgrade and next_upgrade.name)
+
+  local player = game.get_player(event.player_index) --[[@as LuaPlayer]]
+  if add_upgrade_filter(cursor, prototype.name, to) then
+    player.play_sound{path = "utility/smart_pipette"}
+  else
+    util.cursor_notification(player, { "message.bpt-unable-to-add-filter" }, "utility/cannot_build")
+  end
+end)
+
+local lib = {}
+
+function lib.cache_downgrades()
+  local downgrades = {}
```

Comment:
Could you add a type annotation to this?

Reply:
Absolutely